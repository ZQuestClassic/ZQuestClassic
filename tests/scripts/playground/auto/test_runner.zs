namespace Test
{
	void Init(int dmap = 5, int screen = 15)
	{
		printf("[Test] started: dmap %d screen %d\n", dmap, screen);
		if (dmap != 5 || screen != 15)
			Player->Warp(dmap, screen);
	}

	void AssertEqual<T>(T actual, T expected)
	{
		if (expected != actual)
		{
			printf("[Test] failed assertion: expected %d but got %d\n", expected, actual);
			Game->Exit();
		}
	}

	void AssertEqual<T>(T actual, T expected, char32[] error)
	{
		if (expected != actual)
		{
			printf("[Test] failed assertion: expected %d but got %d\n", expected, actual);
			printf("[Test] %s\n", error);
			Game->Exit();
		}
	}

	void Assert<T>(T actual)
	{
		if (!actual)
			Fail("");
	}

	void Assert<T>(T actual, char32[] error)
	{
		if (!actual)
			Fail(error);
	}
	
	void Fail(char32[] error)
	{
		if (strlen(error) > 0)
			printf("[Test] failed test: %s\n", error);
		else
			printf("[Test] failed test\n");
		Game->Exit();
	}
}

screendata script TestRunner
{
	void run()
	{
		char32 name[100];
		Game->GetSaveName(name);
		printf("[Test] Running %s\n", name);

		int script_id = Game->GetGenericScript(name);
		if (script_id == -1)
		{
			printf("[Test] failed to start - could not find test script\n");
			Game->Exit();
		}

		genericdata gd = Game->LoadGenericData(script_id);
		gd->Running = true;
		while (gd->Running)
		{
			Waitframe();
		}

		printf("[Test] done\n");
		Game->Exit();
	}
}
