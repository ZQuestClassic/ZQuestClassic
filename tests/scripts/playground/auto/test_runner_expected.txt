stderr:

Compiling 'test_runner.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
Pass 4: Analyzing Code
Pass 5: Checking code paths
Pass 6: Generating object code
Pass 7: Assembling
Success!
Compile finished with exit code '0' (success)

stdout:

{
  "success": true,
  "diagnostics": [],
  "metadata": {
    "currentFileSymbols": 8,
    "symbols": 18,
    "identifiers": 40,
    "elided": true
  }
}

zasm:

PUSHARGSV 0 0.0006; void run() Body Start
SETR D4 SP2
ALLOCATEMEMV D2 100 0
STORE D2 0.0004
LOAD D2 0.0004; InlineFunc[void Game::GetSaveName(char32[])] Params
GETSAVENAME D2; InlineFunc[void Game::GetSaveName(char32[])] Body
ALLOCATEMEMV D2 19 0
STORE D2 0.0003
WRITEPODSTRING D2 "[Test] Running %s\n"
LOAD D2 0.0003; InlineFunc[void printf(char32[])] Params Start
PUSHR D2
LOAD D2 0.0004
PUSHVARGR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
LOAD D3 0.0003
DEALLOCATEMEMR D3
LOAD D2 0.0004; InlineFunc[int Game::GetGenericScript(char32[])] Params
GETGENERICSCRIPT D2; InlineFunc[int Game::GetGenericScript(char32[])] Body
STORE D2 0.0002
LOAD D2 0.0002; if() #2 Test
COMPAREV D2 -1
SETCMP D2 I==
COMPAREV D2 0
GOTOCMP 36 ==; Test 'if'
ALLOCATEMEMV D2 53 0; if() #2 Body Start
STORE D2 0.0001
WRITEPODSTRING D2 "[Test] failed to start - could not find test script\n"
LOAD D2 0.0001; InlineFunc[void printf(char32[])] Params Start
PUSHR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
LOAD D3 0.0001
DEALLOCATEMEMR D3
GAMEEXIT; InlineFunc[void Game::Exit()] Body AND if() #2 Body End
LOAD D2 0.0002; InlineFunc[genericdata Game::LoadGenericData(int)] Params
LOADGENERICDATA D2; InlineFunc[genericdata Game::LoadGenericData(int)] Body Start
SETR D2 REFGENERICDATA; InlineFunc[genericdata Game::LoadGenericData(int)] Body End
STORE D2 0.0001
SETV D2 1
PUSHR D2
LOAD D2 0.0001
POP D3
PUSHR D2
PUSHR D3
POP D2
CASTBOOLI D2
POP REFGENERICDATA
SETR GENDATARUNNING D2
LOAD D2 0.0001; while() #1 Test
PUSHR D2
POP REFGENERICDATA
SETR D2 GENDATARUNNING
COMPAREV D2 0
GOTOCMP 58 ==; Test 'while'
WAITFRAME; InlineFunc[void Waitframe()] Body AND while() #1 Body
GOTO 50; while() #1 End
ALLOCATEMEMV D2 13 0
STORE D2 0
WRITEPODSTRING D2 "[Test] done\n"
LOAD D2 0; InlineFunc[void printf(char32[])] Params Start
PUSHR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
LOAD D3 0
DEALLOCATEMEMR D3
GAMEEXIT; InlineFunc[void Game::Exit()] Body
LOAD D3 0.0004
DEALLOCATEMEMR D3; [Opt:NeverRet] AND void run() Body End


QUIT; void run() Body
